src/Hack.sol

```diff
@@ -628,53 +631,68 @@ contract WinnerTakesAll {
             rounds.push();
         }
     }
 
-    /// @dev Set the reward at a specific round.
-    /// @param _roundIndex The index of the round concerned by the reward.
-    function setRewardsAtRound(uint256 _roundIndex) external payable onlyOwner {
-        require(rounds[_roundIndex].rewards == 0);
-        rounds[_roundIndex].rewards = msg.value;
+    function buyTicketForLottery(uint256 _lotteryRounds) external payable {
+        require(msg.value == 1 ether, "wrong value");
+        rounds[_lotteryRounds].ticketDistribution[msg.sender].push(rounds[_lotteryRounds].ticketNumber);
+        rounds[_lotteryRounds].ticketNumber++;
     }
 
-    /// @dev Allows the participation of a set of addresses for a specific round.
-    /// @param _roundIndex The index of the round concerned.
-    /// @param _recipients The set of addresses allowed to participate.
-    function setRewardsAtRoundFor(uint256 _roundIndex, address[] calldata _recipients) external onlyOwner {
-        for (uint256 i; i < _recipients.length; i++) {
-            rounds[_roundIndex].isAllowed[_recipients[i]] = true;
-        }
+    /// @dev Set the reward at a specific round.
+    /// @param _lotteryIndex The index of the round concerned by the reward.
+    function setRewardsAtRound(uint256 _lotteryIndex) external payable onlyOwner {
+        require(rounds[_lotteryIndex].rewards == 0);
+        rounds[_lotteryIndex].rewards = msg.value;
     }
 
-    /// @dev Checks if an address can participate in this round.
-    /// @param _roundIndex The index of the round to be checked.
-    /// @param _recipient The address whose authorization is to be checked.
-    function isAllowedAt(uint256 _roundIndex, address _recipient) external view returns (bool) {
-        return rounds[_roundIndex].isAllowed[_recipient];
+    /// @dev Allows the participation of a set of addresses for a specific round.
+    /// @param _lotteryIndex The index of the round concerned.
+    /// @param winningNumber The set of addresses allowed to participate.
+    function setWinningNumberAtRound(uint256 _lotteryIndex, uint256 winningNumber) external onlyOwner {
+        require(winningNumber < rounds[_lotteryIndex].ticketNumber);
+        rounds[_lotteryIndex].winningNumber = winningNumber;
     }
 
     /// @dev Withdraws rewards of a round.
-    /// @param _roundIndex The index of the round concerned.
-    function withdrawRewards(uint256 _roundIndex) external {
-        require(rounds[_roundIndex].isAllowed[msg.sender]);
-        uint256 amount = rounds[_roundIndex].rewards;
-        rounds[_roundIndex].rewards = 0;
-        (bool success,) = msg.sender.call{value: amount}("");
-        require(success, "Transfer failed");
+    /// @param _lotteryIndex The index of the round concerned.
+    function withdrawRewards(uint256 _lotteryIndex) external {
+        uint256[] memory numbers = rounds[_lotteryIndex].ticketDistribution[msg.sender];
+        uint256 winningTicket = rounds[_lotteryIndex].winningNumber;
```

Comment:
ticket 0 always wins at the beginning of the game (while `setWinningNumberAtRound` has not been called. We can start the ticket numbers at 1 and check in the function `withdrawRewards` that the winning number is not zero.